src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs

```diff
@@ -150,22 +195,24 @@ public async Task Handle(RebuildCapabilityStatement notification, CancellationTo
 
         public override async Task<ResourceElement> GetMetadata(CancellationToken cancellationToken = default)
         {
-            await _metadataSemaphore.WaitAsync(cancellationToken);
-            try
+            if (_metadata == null)
             {
-                if (_metadata == null)
-                {
-                    _ = await GetCapabilityStatementOnStartup(cancellationToken);
+                _ = await GetCapabilityStatementOnStartup(cancellationToken);
 
+                await _metadataSemaphore.WaitAsync(cancellationToken);
+                try
+                {
                     _metadata = _builder.Build().ToResourceElement();
+                    return _metadata;
+                }
+                finally
+                {
+                    _metadataSemaphore.Release();
                 }
-
-                return _metadata;
-            }
-            finally
-            {
-                _metadataSemaphore.Release();
             }
+
+            // There is a chance that the BackgroundLoop handler sets metadata to null between when it is checked and returned, but the alternative was leading to deadlocks where the creation of metadata could trigger a rebuild. The rebuild handler had to wait on the metadata semaphore, which wouldn't be released until the metadata could be built. But the metadata builder was waiting on the rebuild handler.
```

Comment:
A pattern here would be to capture the _metadata value in a local variable that gets returned. i.e. 
`if((metadata = _metadata) != null) { return metadata; }`